### ISSUE TITLE
Re-entrance in worker.Mine() call 

### ISSUE THREAD

**Issue by ZenGround0** [filecoin-project/go-filecoin#4065]
**Describe the bug**:
Looks like Worker.Mine() can be called on two concurrent threads which leads to a concurrent map write because of drand.RoundsInInterval use in both threads.  

```
fatal error: concurrent map writes
fatal error: concurrent map writes

goroutine 803 [running]:
runtime.throw(0x20ab87a, 0x15)
        /home/ubuntu/go/src/runtime/panic.go:774 +0x72 fp=0xc075f75138 sp=0xc075f75108 pc=0x52eaa2
runtime.mapassign_fast64(0x1e5bf80, 0xc07549df80, 0x3007, 0xc075390440)
        /home/ubuntu/go/src/runtime/map_fast64.go:176 +0x337 fp=0xc075f75178 sp=0xc075f75138 pc=0x510fe7
github.com/filecoin-project/go-filecoin/internal/pkg/drand.(*GRPC).updateLocalState(...)
        /home/ubuntu/go-filecoin/internal/pkg/drand/drand_grpc.go:121
github.com/filecoin-project/go-filecoin/internal/pkg/drand.(*GRPC).ReadEntry(0xc000469080, 0x23ddee0, 0xc075decdc0, 0x3007, 0xc07b063de0, 0x0, 0x0)
        /home/ubuntu/go-filecoin/internal/pkg/drand/drand_grpc.go:108 +0x339 fp=0xc075f75258 sp=0xc075f75178 pc=0xc30d89
github.com/filecoin-project/go-filecoin/internal/pkg/drand.(*GRPC).RoundsInInterval(0xc000469080, 0x23ddee0, 0xc075decdc0, 0x0, 0xed63e56a4, 0x33e69e0, 0x0, 0xed6422dd3, 0x33e69e0, 0xc075decdc0, ...)
        /home/ubuntu/go-filecoin/internal/pkg/drand/drand_grpc.go:261 +0x1d3 fp=0xc075f75310 sp=0xc075f75258 pc=0xc31e33
github.com/filecoin-project/go-filecoin/internal/pkg/mining.(*DefaultWorker).drandEntriesForEpoch(0xc00046a600, 0x23ddee0, 0xc075decdc0, 0xc000010530, 0x1, 0x1, 0xc0754de8f0, 0x1, 0x1, 0x2753, ...)
        /home/ubuntu/go-filecoin/internal/pkg/mining/worker.go:364 +0x1bf fp=0xc075f75400 sp=0xc075f75310 pc=0xe9f62f
github.com/filecoin-project/go-filecoin/internal/pkg/mining.(*DefaultWorker).Mine(0xc00046a600, 0x23ddee0, 0xc075decdc0, 0xc000010530, 0x1, 0x1, 0xc0754de8f0, 0x1, 0x1, 0x2753, ...)
        /home/ubuntu/go-filecoin/internal/pkg/mining/worker.go:227 +0x659 fp=0xc075f75f10 sp=0xc075f75400 pc=0xe9cee9
github.com/filecoin-project/go-filecoin/internal/pkg/mining.(*timingScheduler).mineLoop.func1(0xc000475cc0, 0xc000010530, 0x1, 0x1, 0xc0754de8f0, 0x1, 0x1, 0x2753, 0xc0754a1800, 0xc07533b400, ...)
        /home/ubuntu/go-filecoin/internal/pkg/mining/scheduler.go:98 +0x8f fp=0xc075f75f80 sp=0xc075f75f10 pc=0xe9febf
runtime.goexit()
        /home/ubuntu/go/src/runtime/asm_amd64.s:1357 +0x1 fp=0xc075f75f88 sp=0xc075f75f80 pc=0x55eaf1
created by github.com/filecoin-project/go-filecoin/internal/pkg/mining.(*timingScheduler).mineLoop
        /home/ubuntu/go-filecoin/internal/pkg/mining/scheduler.go:97 +0x397

```

and 

```
goroutine 766 [running]:
runtime.throw(0x20ab87a, 0x15)
        /home/ubuntu/go/src/runtime/panic.go:774 +0x72 fp=0xc0780e1138 sp=0xc0780e1108 pc=0x52eaa2
runtime.mapassign_fast64(0x1e5bf80, 0xc07549df80, 0x3007, 0xc075390440)
        /home/ubuntu/go/src/runtime/map_fast64.go:176 +0x337 fp=0xc0780e1178 sp=0xc0780e1138 pc=0x510fe7
github.com/filecoin-project/go-filecoin/internal/pkg/drand.(*GRPC).updateLocalState(...)
        /home/ubuntu/go-filecoin/internal/pkg/drand/drand_grpc.go:121
github.com/filecoin-project/go-filecoin/internal/pkg/drand.(*GRPC).ReadEntry(0xc000469080, 0x23ddee0, 0xc077e70c00, 0x3007, 0xc0775179c0, 0x0, 0x0)
        /home/ubuntu/go-filecoin/internal/pkg/drand/drand_grpc.go:108 +0x339 fp=0xc0780e1258 sp=0xc0780e1178 pc=0xc30d89
github.com/filecoin-project/go-filecoin/internal/pkg/drand.(*GRPC).RoundsInInterval(0xc000469080, 0x23ddee0, 0xc077e70c00, 0x0, 0xed63e56a4, 0x33e69e0, 0x0, 0xed6422dec, 0x33e69e0, 0xc077e70c00, ...)
        /home/ubuntu/go-filecoin/internal/pkg/drand/drand_grpc.go:261 +0x1d3 fp=0xc0780e1310 sp=0xc0780e1258 pc=0xc31e33
github.com/filecoin-project/go-filecoin/internal/pkg/mining.(*DefaultWorker).drandEntriesForEpoch(0xc00046a600, 0x23ddee0, 0xc077e70c00, 0xc000010530, 0x1, 0x1, 0xc0754de8f0, 0x1, 0x1, 0x2754, ...)
        /home/ubuntu/go-filecoin/internal/pkg/mining/worker.go:364 +0x1bf fp=0xc0780e1400 sp=0xc0780e1310 pc=0xe9f62f
github.com/filecoin-project/go-filecoin/internal/pkg/mining.(*DefaultWorker).Mine(0xc00046a600, 0x23ddee0, 0xc077e70c00, 0xc000010530, 0x1, 0x1, 0xc0754de8f0, 0x1, 0x1, 0x2754, ...)
        /home/ubuntu/go-filecoin/internal/pkg/mining/worker.go:227 +0x659 fp=0xc0780e1f10 sp=0xc0780e1400 pc=0xe9cee9
github.com/filecoin-project/go-filecoin/internal/pkg/mining.(*timingScheduler).mineLoop.func1(0xc000475cc0, 0xc000010530, 0x1, 0x1, 0xc0754de8f0, 0x1, 0x1, 0x2754, 0xc0754a1800, 0xc07533b400, ...)
        /home/ubuntu/go-filecoin/internal/pkg/mining/scheduler.go:98 +0x8f fp=0xc0780e1f80 sp=0xc0780e1f10 pc=0xe9febf
runtime.goexit()
        /home/ubuntu/go/src/runtime/asm_amd64.s:1357 +0x1 fp=0xc0780e1f88 sp=0xc0780e1f80 pc=0x55eaf1
created by github.com/filecoin-project/go-filecoin/internal/pkg/mining.(*timingScheduler).mineLoop
        /home/ubuntu/go-filecoin/internal/pkg/mining/scheduler.go:97 +0x397

```

Note that block authoring shouldn't hit the `updateLocalState` function of drand at all once we get rid of skipped round handling which will fix the immediate issue + added contention when we mine and sync concurrently.  The underlying surprise that these goroutines run concurrently should be investigated.

**Expected behavior**:
previous worker.Mine() methods canceled before new worker.Mine called.

**Diagnostic information:**

- Filecoin Version:  `d79111acc7396ebc64128238128033b0173c047c`
- Filecoin Inspect Output:  N/A daemon crashed
- Initialization Command:  N/A not a network issue

**Comment by anorth** [comment]
I have seen a very similar crash now at commit 5aff4d3b47a9f5c1f7734aeb1564283da074909f. 

Log file: https://gist.github.com/anorth/a518d63c6b3aca3b7fffec712b8347b7

The crash log has one _one_ goroutine in `drand.(*GRPC).updateLocalState(...)`, but has many in `mining.(*DefaultWorker).drandEntriesForEpoch`

updateLocalState: 1
transport.(*Stream).waitOnHeader: 4
bls12-381.(*KyberG2).Hash: 1
context.propagateCancel inside grpc.NewClientStream: 1

Even though only 1 appears to be in updateLocalState, this might be because the dump can't get a true snapshot across multiple threads, and in fact there was more than one.

The context cancellation doesn't seem to be working. GRPC requests to drand seem to be slooow.

**Comment by anorth** [comment]
The DRAND code has `context.TODO()` on all the endpoints :-(

**Comment by anorth** [comment]
We also ignore the context and it's cancellation in `drand.GRPC` and I think that's the path into this concurrent modification. I think that checking the context explicitly will mostly resolve it, though not robustly. We should also rewrite the mining loop to be completely synchronous. I had some WIP for this earlier related to rush catchup mining in `anorth/catchupmine-2`.

**Comment by anorth** [comment]
Closed by #4139 and $4146
